### ISSUE TITLE
Oddness with JSC_UNSAFE_THIS warning

### ISSUE THREAD

**Issue by juj** [google/closure-compiler#3547]
```js
// ==ClosureCompiler==
// @output_file_name default.js
// @compilation_level ADVANCED_OPTIMIZATIONS
// ==/ClosureCompiler==

var FS = {};

FS.FSNode = /** @constructor */ function(parent, name, mode, rdev) {
 if (!parent) {
  parent = this;
 }
 this.parent = parent;
 this.mount = parent.mount;
 this.mounted = null;
 this.name = name;
 this.mode = mode;
 this.node_ops = {};
 this.stream_ops = {};
 this.rdev = rdev;
};
```
gives
```
JSC_UNSAFE_THIS: dangerous use of 'this' in static method FS.FSNode at line 5 character 11
  parent = this;
           ^
JSC_UNSAFE_THIS: dangerous use of 'this' in static method FS.FSNode at line 7 character 1
 this.parent = parent;
 ^
JSC_UNSAFE_THIS: dangerous use of 'this' in static method FS.FSNode at line 8 character 1
 this.mount = parent.mount;
 ^
JSC_UNSAFE_THIS: dangerous use of 'this' in static method FS.FSNode at line 9 character 1
 this.mounted = null;
 ^
JSC_UNSAFE_THIS: dangerous use of 'this' in static method FS.FSNode at line 10 character 1
 this.name = name;
 ^
JSC_UNSAFE_THIS: dangerous use of 'this' in static method FS.FSNode at line 11 character 1
 this.mode = mode;
 ^
JSC_UNSAFE_THIS: dangerous use of 'this' in static method FS.FSNode at line 12 character 1
 this.node_ops = {};
 ^
JSC_UNSAFE_THIS: dangerous use of 'this' in static method FS.FSNode at line 13 character 1
 this.stream_ops = {};
 ^
JSC_UNSAFE_THIS: dangerous use of 'this' in static method FS.FSNode at line 14 character 1
 this.rdev = rdev;
 ^
```

However
```js
// ==ClosureCompiler==
// @output_file_name default.js
// @compilation_level ADVANCED_OPTIMIZATIONS
// ==/ClosureCompiler==

var FS = {};

var FSNode = /** @constructor */ function(parent, name, mode, rdev) {
 if (!parent) {
  parent = this;
 }
 this.parent = parent;
 this.mount = parent.mount;
 this.mounted = null;
 this.name = name;
 this.mode = mode;
 this.node_ops = {};
 this.stream_ops = {};
 this.rdev = rdev;
};

FS.FSNode = FSNode;
```
completes without warnings. Should there be some functional difference with these two that I'm not catching?

**Comment by rishipal** [comment]
Created a smaller repro for this -

- "var a = {}; a.b = function() {this.c};", // UNSAFE_THIS warning
- "var a = {}; var b = function() { this.c;}; a.b=b;" // no warning

**Comment by rishipal** [comment]
This warning is issued by the CollapseProperties pass to say whether it encountered an unsafe this while collapsing properties. 
The code you have posted issues a different warning - [JSC_USED_GLOBAL_THIS] dangerous use of the global this object. 

This is WAI. 

https://closure-compiler-debugger.appspot.com/#input0%3Dvar%2520FS%2520%253D%2520%257B%257D%253B%250A%250Avar%2520FSNode%2520%253D%2520%252F**%2520constructor%2520*%252F%2520function()%2520%257B%250A%2520%2520return%2520this.c%253B%250A%257D%253B%250A%250AFS.FSNode%2520%253D%2520FSNode%253B%26input1%26conformanceConfig%26externs%26refasterjs-template%26ENABLE_ALL_DIAGNOSTIC_GROUPS%3Dtrue%26CHECK_TYPES%3Dtrue%26REWRITE_MODULES_BEFORE_TYPECHECKING%3Dtrue%26CHECK_GLOBAL_NAMES%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue
